src/state/PollingSubscription.ts

```diff
@@ -0,0 +1,129 @@
+import { Schema } from '../resource';
+import { Subscription, SubscriptionInit } from './SubscriptionManager';
+
+function min(iter: IterableIterator<number>) {
+  let ret = Infinity;
+  for (const n of iter) {
+    if (n < ret) ret = n;
+  }
+  return ret;
+}
+
+/**
+ * PollingSubscription keeps a given resource updated by
+ * dispatching a fetch at a rate equal to the minimum update
+ * interval requested.
+ */
+export default class PollingSubscription implements Subscription {
+  protected readonly schema: Schema;
+  protected readonly fetch: () => Promise<any>;
+  protected readonly url: string;
+  protected frequency: number;
+  protected frequencyHistogram: Map<number, number> = new Map();
+  protected dispatch: React.Dispatch<any>;
+  protected intervalId?: NodeJS.Timeout;
+  protected lastIntervalId?: NodeJS.Timeout;
+
+  constructor(
+    { url, schema, fetch, frequency }: SubscriptionInit,
+    dispatch: React.Dispatch<any>,
+  ) {
+    if (frequency === undefined)
+      throw new Error('frequency needed for polling subscription');
+    this.schema = schema;
+    this.fetch = fetch;
+    this.frequency = frequency;
+    this.url = url;
+    this.frequencyHistogram.set(this.frequency, 1);
+    this.dispatch = dispatch;
+    this.run();
+  }
+
+  /** Subscribe to a frequency */
+  add(frequency?: number) {
+    if (frequency === undefined) return;
+    if (this.frequencyHistogram.has(frequency)) {
+      this.frequencyHistogram.set(
+        frequency,
+        (this.frequencyHistogram.get(frequency) as number) + 1,
+      );
+    } else {
+      this.frequencyHistogram.set(frequency, 1);
+
+      // new min so restart service
+      if (frequency < this.frequency) {
+        this.frequency = frequency;
+        this.run();
+      }
+    }
+  }
+
+  /** Unsubscribe from a frequency */
+  remove(frequency?: number) {
+    if (frequency === undefined) return false;
+    if (this.frequencyHistogram.has(frequency)) {
+      this.frequencyHistogram.set(
+        frequency,
+        (this.frequencyHistogram.get(frequency) as number) - 1,
+      );
+      if ((this.frequencyHistogram.get(frequency) as number) < 1) {
+        this.frequencyHistogram.delete(frequency);
+
+        // this was the min, so find the next size
+        if (frequency <= this.frequency) {
+          if (this.frequencyHistogram.size === 0) return true;
+          this.frequency = min(this.frequencyHistogram.keys());
+          this.run();
+        }
+      }
+    } else if (process.env.NODE_ENV !== 'production') {
```

Comment:
What happens here if a user of `rest-hooks` doesn't have a DefinePlugin with `process.env`?

Reply:
It depends on which build they are consuming. The unpkg one gets this compiled out. The es6 module version will keep it in assuming they are using a standard webpack or rollup setup.

If these assumptions turn out to be bad I'm sure there will be issues created to think about future direction; but until those people pop up I can't really assume anything about their use case.